pkg/kubelet/images/image_manager.go

```diff
@@ -224,3 +372,97 @@ func applyDefaultImageTag(image string) (string, error) {
 	}
 	return image, nil
 }
+
+// isEnsuredBySecret - returns two bools and error
+// 1. pulledBySecret: true if it is in ensured secret pulled image list,
+// the list will clean the image after PullImageSecretRecheckPeriod
+// 2. ensuredBySecret: true if the secret for an auth used to pull an
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image.
+func (m *imageManager) isEnsuredBySecret(imageRef string, image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (bool, bool, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) && m.pullImageSecretRecheck {
+		m.refreshCache()
+	}
+	if imageRef == "" {
+		return false, false, errors.New("imageRef is empty")
+	}
+
+	// if the image is in the ensured secret pulled image list, it is pulled by secret
+	pulledBySecret := m.ensureSecretPulledImages[imageRef] != nil
+
+	img := image.Image
+	repoToPull, _, _, err := parsers.ParseImageName(img)
+	if err != nil {
+		return pulledBySecret, false, err
+	}
+
+	if m.keyring == nil {
+		return pulledBySecret, false, nil
+	}
+	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, *m.keyring)
+	if err != nil {
+		return pulledBySecret, false, err
+	}
+
+	creds, withCredentials := keyring.Lookup(repoToPull)
+	if !withCredentials {
+		return pulledBySecret, false, nil
+	}
+	// if the pullImageSecretRecheckPeriod is 0, we don't need to check the secret again
+	if m.pullImageSecretRecheckPeriod.Duration == 0 {
+		return pulledBySecret, true, nil
```

Comment:
I don't know if this is right. If the duration is 0, then we don't delete the entry in the cache in refreshCache.

If pulledBySecret is true, then it's present in the cache, but not necessarily having matching auths (you check the auths below)
if it's false, then it's either the first time we have pulled the image, or the image has had it's auths cleaned up from the cache, and we should recheck.

Reply:
Yeah, I agree with you here, I have already added that condition and this piece of code is from the previous implementation which I felt could be removed. Thanks for pointing out !